posix-list-dir_test.go

```diff
@@ -0,0 +1,114 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+
+	. "gopkg.in/check.v1"
+)
+
+type PosixTestSuite struct {
+	dir   string
+	names []string
+}
+
+var _ = Suite(&PosixTestSuite{})
+
+func (s *PosixTestSuite) TestReadDir0(c *C) {
+	// Check non existent directory.
+	_, err := readDir("/tmp/non-existent-directory")
+	c.Assert(err, Equals, errFileNotFound)
+
+	// Check if file is given.
+	_, err = readDir("/etc/issue/mydir")
+	c.Assert(err, Equals, errFileNotFound)
+
+	// Check if permission denied.
+	_, err = readDir("/proc/1/fd")
+	c.Assert(err, Not(Equals), nil)
+}
+
+func (s *PosixTestSuite) SetupTestReadDir1(c *C) {
+	dir, err := ioutil.TempDir("", "minio-posix-list-dir")
+	c.Assert(err, Equals, nil)
+
+	s.dir = dir
+}
+
+func (s *PosixTestSuite) TeardownTestReadDir1(c *C) {
+	os.RemoveAll(s.dir)
+}
+
+// Test to read empty directory.
+func (s *PosixTestSuite) TestReadDir1(c *C) {
```

Comment:
Instead of multiple test functions, its better to generate a array of struct with all inputs and expected outputs and iterate through them, here is an example https://github.com/minio/minio/blob/master/object-api-multipart_test.go#L132 . This allows for adding more cases easily.